### PR TITLE
committed

### PR DESCRIPTION
**Pull request by excellencetechnologies**
Pull Request after implemented cron.

### CODE REVIEW

File: mods/db.js [hunk 1]
```diff
@@ -1,32 +1,64 @@
...
+        status: {type: String, required: true, unique: true},
+        cron_running_time: {type: String, required: true, unique: true}
+    });
+    var app_urls = mongoose.model('AppUrls', app_url_schema);
+    mongoose.connect(config.DB_URL, function (err, db) {
+
+
+        var categoryListSchema = mongoose.Schema({}, {
+            strict: false,
+            collection: 'categoryList'
```
Comment (upexcel) [inline mods/db.js]: categoryListCache

File: mods/db.js [hunk 2]
```diff
@@ -1,32 +1,64 @@
...
+        });
+        var CollectioncategoryList = conn.model('categoryList', categoryListSchema);
+        var homeSchema = mongoose.Schema({}, {
+            strict: false,
```
Comment (upexcel) [inline mods/db.js]: strict: false remove and put proper schema

File: mods/db.js [hunk 3]
```diff
@@ -1,32 +1,64 @@
...
+            collection: 'homeSlider'
+        });
+        var homeSlider = conn.model('homeSlider', homeSchema);
+        var homeProductSchema = mongoose.Schema({}, {
+            strict: false,
```
Comment (upexcel) [inline mods/db.js]: define in a seperate file called schema.js

File: mods/db.js [hunk 4]
```diff
@@ -1,32 +1,64 @@
...
+            collection: 'homeProducts'
+        });
+        var homeProducts = conn.model('homeProducts', homeProductSchema);
+        app_urls.find({}, {APP_ID: 1, _id: 0}, function (err, value) {
```
Comment (upexcel) [inline mods/db.js]: put this in a service,

file name: magento.js
function: getActiveInstallations()

File: mods/db.js [hunk 5]
```diff
@@ -1,32 +1,64 @@
...
+            if (err) {
+                console.log(err);
+            } else if (!value) {
+                console.log(value);
+            } else {
+                for (i = 0; i < value.length; i++) {
```
Comment (upexcel) [inline mods/db.js]: right now all active magento are in a for loop. this is wrong. 
we should process one and then when its finished next

File: mods/db.js [hunk 6]
```diff
@@ -1,32 +1,64 @@
...
+                    app_id = value[i].get('APP_ID');
+                    cron(app_urls, CollectioncategoryList, homeSlider, homeProducts, app_id);
```
Comment (upexcel) [inline mods/db.js]: rename cron function to processStore() and don't send schema individually

Comment (upexcel) [inline mods/db.js]: rename cron function to processStore() and don't send schema individually

File: package.json [hunk 7]
```diff
@@ -32,12 +34,14 @@
...
     "redis": "^2.6.2",
     "request": "^2.74.0",
+    "serve-static": "^1.11.1",
     "sharp": "^0.16.2",
     "static-favicon": "~1.0.0",
     "underscore": "^1.8.3",
     "url": "^0.11.0"
   },
   "devDependencies": {
+    "babel-cli": "^6.18.0",
```
Comment (upexcel) [inline package.json]: remove this

File: service/cron.js [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+var jstz = require('jstz');
+var timezone = jstz.determine().name();
+require('./category');
+require('./home');
+
+cron = function (AppUrls, CollectioncategoryList, homeSliderList, homeProductsList, app_id) {
+
+    // pattern for crone  after 5 min '*/5 * * * *'
+    new CronJob('* * * * * *', function () {
+        AppUrls.findOne({APP_ID: app_id}, function (err, user) {
```
Comment (upexcel) [inline service/cron.js]: dont do findOne again. we already have done find above. directly pass the object in this function

File: service/cron.js [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+            if (err) {
+                console.log(err);
+            } else if (!user) {
+                console.log(user);
+            } else {
+                cron_running_time = user.cron_running_time;
+                var current_time = moment().tz('Asia/Calcutta').format('HH:mm:ss ZZ'); //13:56:34 +0530
```
Comment (upexcel) [inline service/cron.js]: when we are comparing time we just need to compare hours and minutes

File: service/cron.js [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+                var format = 'HH:mm:ss ZZ';
+                var cron_running_time_with_IST = moment(cron_running_time, format).tz('Asia/Calcutta').format(format);
+                if (current_time == cron_running_time_with_IST) {
+                    console.log('You will see this message every minute');
+                    var categoryListDB = CollectioncategoryList;
+                    categoryListDB.findOne({
```
Comment (upexcel) [inline service/cron.js]: create service preFeatch.js 
and make different function fetchCategory , fetchHomeProducts, ....

File: service/cron.js [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+                        cache: 0
+                    }, function (error, result) {
```
Comment (upexcel) [inline service/cron.js]: basically right now category list, home products, home slider all are fetching together. 
instead it should be one after another.

to do this in app_urls collection add following

1. prefetch_status: 'FINISHED', 'RUNNING'
2. prefetch_last_run: time when last finished if status is finished or time when started if status running 
3. prefetch_current_status:  'CATEGORY' , 'HOME_SLIDER', 'HOME_PRODUCT'  or blank

File: service/cron.js [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+                            var req = {headers: {app_id: config.APP_ID},
+                                body: {mobile_width: '300'},
+                                URL: config.URL
+                            };
+                            homeProducts(req, function (body) {
+                                if (body.status == 0) {
+                                } else {
+
+                                    var allData = body.msg;
+                                    for (var a = allData.length - 1; a >= 0; a--) {
```
Comment (upexcel) [inline service/cron.js]: use lodash

File: package.json (hunk 7 above)
Comment (excellencetechnologies) [inline package.json]: Fixed

File: service/cron.js (hunk 9 above)
Comment (excellencetechnologies) [inline service/cron.js]: Fixed

File: mods/db.js (hunk 1 above)
Comment (excellencetechnologies) [inline mods/db.js]: Fixed

File: mods/db.js (hunk 2 above)
Comment (excellencetechnologies) [inline mods/db.js]: Fixed

File: mods/db.js (hunk 3 above)
Comment (excellencetechnologies) [inline mods/db.js]: Fixed

File: service/cron.js (hunk 12 above)
Comment (excellencetechnologies) [inline service/cron.js]: Fixed

File: mods/db.js (hunk 4 above)
Comment (excellencetechnologies) [inline mods/db.js]: Fixed

File: mods/db.js (hunk 6 above)
Comment (excellencetechnologies) [inline mods/db.js]: Fixed

Comment (excellencetechnologies) [inline mods/db.js]: Fixed

**Pull request merged**